### ISSUE TITLE
a fsnotify-related risk

### ISSUE THREAD

**Issue by mars1024** [cri-o/ocicni#42]
Ocicni now is using fsnotify to load cni config files real-time, but fsnotify has a risk, if we remove cni config path `/etc/cni/net.d` and recreate it, any changes of this directory will not be detected by fsnotify any more. In other words, ocicni will not work properly after cni config path was removed.

**Comment by mars1024** [comment]
@dcbw @mrunalp  PTAL ~

**Comment by mrunalp** [comment]
Is that something that can be fixed in fsnotify upstream?

**Comment by mars1024** [comment]
> Is that something that can be fixed in fsnotify upstream?

I don't think so, because fsnotify is using inotify in linux, and inotify is using inode. Maybe we need some changes in ocicni.

**Comment by kwilczynski** [comment]
@mars1024, what use case would favour removing the entire directory over files within it?

Do you also want to cover the removal of the **/etc/cni** aside from **/etc/cni/net.d**? Monitoring /etc to look for /etc/cni deletion can be done, but it would be a bit of an overkill, I think. While keeping the scope to /etc/cni only might be more feasible, I suppose.

**Comment by saschagrunert** [comment]
Closing as won't fix. While the fsnotify watch is technically lost if the config directory is removed and recreated, this is an edge case that has not caused real problems in 6+ years of production use. In normal CRI-O operation the config directory is never deleted.
